### PR TITLE
feat: Adds PKCE authentication for GitLab closes #5236

### LINKED ISSUE #5236: Add PKCE authentication for gitlab
**Is your feature request related to a problem? Please describe.**

Gitlab will be deprecating implicit authentication. It would be good to add PKCE authentication which is what they will be supporting going forwards.

**Describe the solution you'd like**

For NetlifyCms to support PKCE authentication for gitlab

**Describe alternatives you've considered**

Running some external oauth client. 

**Additional context**

I copied the existing implicit oauth implimentation and modified it to get pkce working. See here: [https://github.com/something-something-something/netlify-cms/commit/0ad28b7d6ed8f0a46f41dffe4e38dd726da34bba](https://github.com/something-something-something/netlify-cms/commit/0ad28b7d6ed8f0a46f41dffe4e38dd726da34bba)

If I clean it up would you consider accepting a pull request for this?

see also https://github.com/netlify/netlify-cms/pull/4878

### PR DESCRIPTION
**Pull request by something-something-something**
**Summary**

closes #5236
This adds PKCE authentication for GitLab as the existing authentication method (implicit)  is going to be deprecated. 
see https://gitlab.com/gitlab-org/gitlab/-/issues/288516 and https://docs.gitlab.com/ce/api/oauth2.html#authorization-code-with-proof-key-for-code-exchange-pkce

Configuration for use is as follows:

```
backend:
  name: gitlab
  repo: user/repo
  auth_type: pkce
  app_id: someapid898909439034904390490
```

I copied the existing implicit oauth implementation and changed it to perform pkce authentication instead. 

As such some functions and other code like createNonce()  and  validateNonce() are the same as the ones in the implicit implementation and should be moved to a separate file. Where should they go?

**Test plan**

None yet. Where do tests go for authentication?

**A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Review by erezrokah — CHANGES_REQUESTED** [review]
Thanks , can we add some information about setting this up in the docs https://www.netlifycms.org/docs/gitlab-backend/#client-side-implicit-grant-gitlab
and maybe add a note about the deprecation of implicit auth?

You can create a new file called `utils.js` file under `netlify-cms-lib-auth/src` and move shared code to it.

For tests, we [pre-recorded browser traffic](https://github.com/netlify/netlify-cms/tree/master/cypress/fixtures) to test integrations with different providers.

I don't think we test implicit auth though.

**Comment by something-something-something** [comment]
Okay I moved the shared functions to utils.js and added docs. 

For docs should implicit just say something like:

Follow the instructions for PKCE but replace 
```yaml
  auth_type: pkce
```
with
```yaml
  auth_type: implicit
```

> I don't think we test implicit auth though.

Is it okay if I don't write test for pkce?

**Review by erezrokah — APPROVED** [review]
This is great @something-something-something.

I made some changes in https://github.com/netlify/netlify-cms/pull/5239/commits/9c7dbb600986cc87a5995e2fad0ec7c4528a3bd4 but they are mostly cosmetic.

If you're happy with https://github.com/netlify/netlify-cms/pull/5239/commits/9c7dbb600986cc87a5995e2fad0ec7c4528a3bd4 I think this is ready.

Since adding tests for this is quite tricky, I don't think this should block the PR.

File: packages/netlify-cms-backend-gitlab/src/AuthenticationPage.js [hunk 1]
```diff
@@ -34,11 +42,12 @@ export default class GitLabAuthenticationPage extends React.Component {
       app_id = '',
     } = this.props.config.backend;
 
-    if (authType === 'implicit') {
-      this.auth = new ImplicitAuthenticator({
+    if (clientSideAuthenticators[authType]) {
```
Comment (erezrokah) [inline packages/netlify-cms-backend-gitlab/src/AuthenticationPage.js:45]: Avoids some code duplication

File: packages/netlify-cms-lib-auth/src/pkce-oauth.js [hunk 2]
```diff
@@ -3,33 +3,49 @@ import trimEnd from 'lodash/trimEnd';
...
-    new Uint8Array(await window.crypto.subtle.digest('SHA-256', new TextEncoder().encode(text))),
-  )
-    .map(el => {
-      return String.fromCharCode(el);
-    })
-    .join('');
+  const encoder = new TextEncoder();
+  const data = encoder.encode(text);
+  const digest = await window.crypto.subtle.digest('SHA-256', data);
+  const sha = String.fromCharCode(...new Uint8Array(digest));
```
Comment (erezrokah) [inline packages/netlify-cms-lib-auth/src/pkce-oauth.js:9]: `Uint8Array` is already an array and `String.fromCharCode` accepts a sequence

File: packages/netlify-cms-lib-auth/src/pkce-oauth.js [hunk 3]
```diff
@@ -3,33 +3,49 @@ import trimEnd from 'lodash/trimEnd';
...
-//based on https://github.com/auth0/auth0-spa-js/blob/9a83f698127eae7da72691b0d4b1b847567687e3/src/utils.ts#L147
+// based on https://github.com/auth0/auth0-spa-js/blob/9a83f698127eae7da72691b0d4b1b847567687e3/src/utils.ts#L147
 function generateVerifierCode() {
-  const randomValuesArr = new Uint8Array(128);
-  return Array.from(window.crypto.getRandomValues(randomValuesArr))
+  // characters that can be used for codeVerifer
+  // excludes _~ as if included would cause an uneven distribution as char.length would no longer be a factor of 256
+  const chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-.';
+  const randomValues = window.crypto.getRandomValues(new Uint8Array(128));
+  return randomValues
```
Comment (erezrokah) [inline packages/netlify-cms-lib-auth/src/pkce-oauth.js:19]: Same, `Uint8Array` is already an array

File: packages/netlify-cms-lib-auth/src/pkce-oauth.js [hunk 4]
```diff
@@ -58,24 +74,17 @@ export default class PkceAuthenticator {
...
-        'code_challenge',
-        btoa(hash)
-          .split('=')[0]
-          .replace(/\+/g, '-')
-          .replace(/\//g, '_'),
-      );
-      //
-      document.location.assign(authURL.href);
-    });
+    const codeVerifier = createCodeVerifier();
```
Comment (erezrokah) [inline packages/netlify-cms-lib-auth/src/pkce-oauth.js:77]: Make it easier to understand the flow

File: packages/netlify-cms-lib-auth/src/pkce-oauth.js [hunk 5]
```diff
@@ -3,33 +3,49 @@ import trimEnd from 'lodash/trimEnd';
...
     .map(val => {
-      // characters that can be used for codeVerifer
-      // excludes _~ as if included would cause an uneven  distribution as char.length would no longer be a factor of 256
-      const chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-.';
       return chars[val % chars.length];
     })
     .join('');
 }
 
-function createOrGetCodeVerifier() {
```
Comment (erezrokah) [inline packages/netlify-cms-lib-auth/src/pkce-oauth.js]: I separated this function to make it more explicit.
Before it seems that creating a code verifier in `completeAuth` could be consider as a valid use case (am I assuming correctly that it is not?)

File: packages/netlify-cms-lib-auth/src/pkce-oauth.js (hunk 3 above)
Comment (something-something-something) [reply packages/netlify-cms-lib-auth/src/pkce-oauth.js:19]: I think You need Array.from here

See the difference in the following (can be pasted into a browser):


The following only produces numbers:

```js
(()=>{
	const chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-.';
	const randomValues = window.crypto.getRandomValues(new Uint8Array(128));
	return randomValues.map((val) => {
		return chars[val % chars.length];
	}).join('');
})();
```

Will produce proper output:
```js
(()=>{
	const chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-.';
	const randomValues = Array.from(window.crypto.getRandomValues(new Uint8Array(128)));
	return randomValues.map((val) => {
		return chars[val % chars.length];
	}).join('');
})();

```

Comment (something-something-something) [reply packages/netlify-cms-lib-auth/src/pkce-oauth.js:19]: The issue is the Uint8Array.map returns an Uint8Array. not a regular array so all the strings are converted to numbers. Strings that can't convert to numbers become 0.

**Comment by something-something-something** [comment]
Thanks! I added back Array.from to generateVerifierCode() as it does in fact need to be there. All the other changes are good with me. I am happy with https://github.com/something-something-something/netlify-cms/commit/aaedd995e669115257f4611ae3aee8d5148cecc2 being merged.

File: packages/netlify-cms-lib-auth/src/pkce-oauth.js (hunk 3 above)
Comment (erezrokah) [reply packages/netlify-cms-lib-auth/src/pkce-oauth.js:19]: Nice one thanks for explaining!

**Pull request merged**